File: compiler/bench/happy.c
// expected: 142
// Happy numbers benchmark - count happy numbers up to 1000, run 500 times

#include <stdio.h>

int sum_digit_squares(int n) {
    int sum = 0;
    while (n) {
        int d = n % 10;
        sum += d * d;
        n /= 10;
    }
    return sum;
}

int happy(int n) {
    for (int i = 0; i < 50; i++) {
        n = sum_digit_squares(n);
        if (n == 1) return 1;
    }
    return 0;
}

int count_happy(int limit) {
    int count = 0;
    for (int i = 1; i < limit; i++) {
        if (happy(i)) count++;
    }
    return count;
}

int main(void) {
    int result = 0;
    for (int i = 0; i < 500; i++) {
        result = count_happy(1000);
    }
    printf("%d\n", result);
    return 0;
}

File: compiler/bench/happy.fs
\ expected: 142
\ Happy numbers benchmark - count happy numbers up to 1000, run 500 times

: sum-digit-squares ( n -- sum )
  0 swap
  begin dup while
    10 /mod swap dup * rot + swap
  repeat drop ;

: happy? ( n -- flag )
  50 0 do
    sum-digit-squares
    dup 1 = if drop -1 unloop exit then
  loop drop 0 ;

: count-happy ( limit -- count )
  0 swap 1 do
    i happy? if 1+ then
  loop ;

: main ( -- )
  0
  500 0 do
    drop 1000 count-happy
  loop
  . cr ;
